Documentation/proposals/202405-agent-daemonset.md

```diff
@@ -0,0 +1,129 @@
+
+We will add a new `mode` field that accepts either `StatefulSet` or `DaemonSet`, with `StatefulSet` being the default. If the DaemonSet mode is activated (`mode: DaemonSet`), all the unrelated fields listed above will not be accepted. In the MVP, we will simply fail the reconciliation if any of those fields are set. We will prevent users from directly switching from a live StatefulSet setup to DaemonSet, because that might break their workload if they forget to unset the unsupported fields. Following up, we will leverage validation rules with [Kubernetes' Common Expression Language (CEL)](https://kubernetes.io/docs/reference/using-api/cel/). Only then, we will allow switching from a live StatefulSet setup to DaemonSet. We already have an issue for CEL [here](https://github.com/prometheus-operator/prometheus-operator/issues/5079).
+
+### 6.2. Node detecting:
+
+As pointed out in [Danny from GMP’s talk](https://www.youtube.com/watch?v=yk2aaAyxgKw), to make Prometheus Agent DaemonSet know which node it’s on, we can use [Kubernetes’ downward API](https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/). In `config-reloader` container, we can mount the node name as an environment variable like this:
+```
+containers:
+- name: config-reloader
+  env:
+  - name: NODE_NAME
+    valueFrom:
+      fieldRef:
+        apiVersion: v1
+        fieldPath: spec.nodeName
+```
+
+### 6.3. Targets filtering for pods (PodMonitor support):
+
+To filter pod targets, Danny’s talk has pointed out the best option is to use field selector:
+```
+kubernetes_sd_configs:
+- role: pod
+  selectors:
+  - role: pod
+    field: spec.nodeName=$NODE_NAME
+```
+We'll go with this option, because it filters targets right at discovery time, and also because Kubernetes API server watch cache indexes pods by node name (as we can see in [Kubernetes codebase](https://github.com/kubernetes/kubernetes/blob/v1.30.0-rc.0/pkg/registry/core/pod/storage/storage.go#L91)).
+
+We've also considered using relabel config that filters pods by `__meta_kubernetes_pod_node_name` label. However, we didn't choose to go with this option because it filters pods only after discovering all the pods from PodMonitor, which increases load on Kubernetes API server.
+
+
+## 7. Action Plan
+
+For the implementation, we’ll do what we detailed in the How section. The common logics between StatefulSet and DaemonSet modes will be extracted into one place. We will have a separate `daemonset.go` for the separate logic of the DaemonSet mode.
+
+For the test, we will have unit tests covering new logic, and integration tests covering the basic user cases, which are:
+* Prometheus Agent DaemonSet cannot be deployed when a similar Prometheus Agent StatefulSet resource is there.
+* Prometheus Agent DaemonSet is created/deleted successfully.
+* Prometheus Agent DaemonSet is installed on the right nodes.
+* Prometheus Agent DaemonSet scales up/down following the scale of nodes.
```

Comment:
After discussing offline, we realized that it would mean testing the DaemonSet functionality offered by Kubernetes which doesn't add any value to the operator validation.